### ISSUE TITLE
(Re)starting machine sometimes results in machine readable failed

### ISSUE THREAD

**Issue by jakirkham** [docker/machine#2494]
Am using Docker Toolbox 1.9.1b with all regularly included version on Mac OS 10.9.5. In particular, am using docker-machine version `0.5.2 ( 0456b9f )`. When (re)starting a machine I frequently, see the following error.

    /usr/local/bin/VBoxManage showvminfo default --machinereadable failed:
    VBoxManage: error: The object is not ready
    VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), callee nsISupports
    VBoxManage: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp

**Comment by askb** [comment]
+1 hitting this issue on Linux / Ubuntu 14.04 - docker-machine version 0.5.3-dev ( 98a1c8c )

```
$ PATH="$PWD/bin:$PATH" bin/docker-machine stop dev9
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), callee nsISupports
VBoxManage: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp
```

**Comment by nathanleclaire** [comment]
Does it do it every time, or just sporadically?

**Comment by jakirkham** [comment]
@nathanleclaire, I initially thought it was sporadically, but not I can reproduce this every time. However, I only see it when the machine is already running. If it is `stopped` and I start it, then this doesn't happen. Could it be a timing issue?

**Comment by askb** [comment]
This could be happening due to a potential race-condition between VM stop's - start's. On my setup running these commands in a loop the issue easily reproducible.

```
$ for i in {1..15}; do PATH="$PWD/bin:$PATH" bin/docker-machine start dev9; PATH="$PWD/bin:$PATH" bin/docker-machine stop dev9"; done
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
(dev9) Stopping VM...
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
(dev9) Stopping VM...
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), component SessionMachine, interface IMachine, callee nsISupports
VBoxManage: error: Context: "COMGETTER(ChipsetType)(&aChipset)" at line 74 of file VBoxManageControlVM.cpp
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), component SessionMachine, interface IMachine, callee nsISupports
VBoxManage: error: Context: "COMGETTER(BandwidthControl)(bwCtrl.asOutParam())" at line 2243 of file VBoxManageInfo.cpp
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
(dev9) Stopping VM...
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), callee nsISupports
VBoxManage: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), callee nsISupports
VBoxManage: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), component ConsoleWrap, interface IConsole, callee nsISupports
VBoxManage: error: Context: "COMGETTER(VRDEServerInfo)(vrdeServerInfo.asOutParam())" at line 2340 of file VBoxManageInfo.cpp
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), callee nsISupports
VBoxManage: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp
(dev9) Starting VM...
^[[BStarted machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
(dev9) Stopping VM...
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), component SessionMachine, interface IMachine, callee nsISupports
VBoxManage: error: Context: "GetSettingsFilePath(bstr.asOutParam())" at line 516 of file VBoxManageInfo.cpp
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), callee nsISupports
VBoxManage: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
(dev9) Stopping VM...
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
(dev9) Stopping VM...
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
/usr/bin/VBoxManage showvminfo dev9 --machinereadable failed:
VBoxManage: error: The object is not ready
VBoxManage: error: Details: code E_ACCESSDENIED (0x80070005), callee nsISupports
VBoxManage: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp
(dev9) Starting VM...
Started machines may have new IP addresses. You may need to re-run the `docker-machine env` command.
(dev9) Stopping VM...
```

```
$ VBoxManage --version
5.0.10r104061
```

**Comment by jakirkham** [comment]
Convinced that you are correct @askb in think this is a race condition. Every time I break this up into stop and start commands everything just works.

**Comment by nathanleclaire** [comment]
VirtualBox is sensitive and will start failing if you throw commands at it too fast.  I'm not sure if anything can be revised in Machine to fix this, except maybe adding sleeps before / after start and stop operations (which is just a band-aid around the problem, it doesn't guarantee it won't happen).

Do you have any ideas?

**Comment by jakirkham** [comment]
So, if I run this in the terminal `docker-machine stop <MACHINE_NAME> ; docker-machine start <MACHINE_NAME>`, it works fine. I have yet to see a failure and have used it since this issue was opened.

Maybe just having restart be a script or `go` code that use a subprocess or call into the API to execute each command and wait on their completion would solve it. As the machine could already be stopped maybe a force option is appropriate for `docker-machine stop` in this case (and possibly `docker-machine start` for other cases) so as not to get an error code when issuing a potentially redundant command.

**Comment by nathanleclaire** [comment]
> Maybe just having restart be a script or go code that use a subprocess or call into the API to execute each command and wait on their completion would solve it.

We do poll the machine to check for state after each command to ensure it finished arriving at the state we desire.

> As the machine could already be stopped maybe a force option is appropriate for docker-machine stop in this case (and possibly docker-machine start for other cases) so as not to get an error code when issuing a potentially redundant command.

We do have `docker-machine kill` which is meant to be more of a "hard stop".

**Comment by jakirkham** [comment]
> We do poll the machine to check for state after each command to ensure it finished arriving at the state we desire.

Can you point me to where this is happening for `docker-machine restart`? Namely, after `stop` and before `start`. Or is this actually a single VirtualBox command?

**Comment by nathanleclaire** [comment]
@jakirkham https://github.com/docker/machine/blob/master/libmachine/host/host.go#L79

**Comment by jakirkham** [comment]
So, I have now seen this happen once when breaking it up into start/stop commands. It could be that this is a VirtualBox bug. The easiest fix would probably just be to add a timed sleep here or increase the time if there is one.

**Comment by jakirkham** [comment]
Also, I have seen this warning when creating new machines. [`0a7406`@libmachine/provision/boot2docker.go#L198-L208]( https://github.com/docker/machine/blob/0a74068b1128a2b7d3a1c0608ee996ddfd796609/libmachine/provision/boot2docker.go#L198-L208 ) Not sure if it is related to the same fundamental problem here. The reason I think it might be is I see this warning and then the next `docker` command works fine.

**Comment by ozlerhakan** [comment]
**Stopping** a machine sporadically gets failure in docker 1.10 on windows 10

Do you want me to create another issue for that? @nathanleclaire 

```
$ docker-machine version
docker-machine.exe version 0.6.0, build e27fb87
$ docker version
Client:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 18:18:11 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 19:55:25 2016
 OS/Arch:      linux/amd64

```
```
$ docker-machine stop local
Stopping "local"...
C:\Program Files\Oracle\VirtualBox\VBoxManage.exe showvminfo local--machinereadable failed:
VBoxManage.exe: error: Code E_ACCESSDENIED (0x80070005) - General access denied error (extended info not available)
VBoxManage.exe: error: Context: "COMGETTER(RemoteUSBDevices)(ComSafeArrayAsOutParam(coll))" at line 2022 of file VBoxManageInfo.cpp
```

**Comment by nathanleclaire** [comment]
Yes, please file a new issue @ozlerhakan

**Comment by Koriit** [comment]
Had the same issue. Changing restart into stop and start resolved the issue.
